File: partial-refund-gives-back-everything/python/audit_refunds.py
"""Catch WooCommerce partial refunds that actually returned the whole charge.

On an order whose PaymentIntent was captured for less than the order total
(a manual capture, a phone order finished outside checkout, a split payment),
WooCommerce computes "amount left to refund" from the order total instead of
the real Stripe amount_captured. Ask for a small partial refund and the
gateway can send Stripe a refund with no amount, or an amount larger than
what is actually left, so Stripe refunds the entire remaining balance.

This script compares, per order, the refund the shop manager intended
(the WooCommerce refund line item) against what Stripe actually refunded on
the matching charge. When Stripe refunded more than intended, it writes an
order note flagging the gap. It never asks Stripe to move money and never
edits amounts. It only reports. Read only by default. Run on a schedule.

Guide: https://www.allanninal.dev/woocommerce/partial-refund-gives-back-everything/
"""
import os
import time
import logging
import stripe
import requests
from requests.auth import HTTPBasicAuth

logging.basicConfig(level=logging.INFO, format="%(asctime)s %(levelname)s %(message)s")
log = logging.getLogger("audit_refunds")

stripe.api_key = os.environ["STRIPE_SECRET_KEY"]
WOO_URL = os.environ["WOO_STORE_URL"].rstrip("/")
AUTH = HTTPBasicAuth(os.environ["WOO_CONSUMER_KEY"], os.environ["WOO_CONSUMER_SECRET"])
LOOKBACK_HOURS = int(os.environ.get("LOOKBACK_HOURS", "72"))
DRY_RUN = os.environ.get("DRY_RUN", "true").lower() == "true"

# Tolerance in cents for rounding noise between the two systems.
TOLERANCE_MINOR = 1


def intent_id_of(order):
    """The saved Stripe PaymentIntent id, from meta _stripe_intent_id or transaction_id."""
    for meta in order.get("meta_data") or []:
        if meta.get("key") == "_stripe_intent_id" and meta.get("value"):
            return meta["value"]
    tid = order.get("transaction_id")
    return tid if tid and (tid.startswith("pi_") or tid.startswith("ch_")) else None


def recently_refunded_orders(lookback_hours):
    """WooCommerce orders that had a refund recorded in the lookback window."""
    since = time.strftime("%Y-%m-%dT%H:%M:%S", time.gmtime(time.time() - lookback_hours * 3600))
    page = 1
    while True:
        r = requests.get(
            f"{WOO_URL}/wp-json/wc/v3/orders",
            params={"status": "refunded,processing,completed", "after": since, "per_page": 50, "page": page},
            auth=AUTH, timeout=30,
        )
        r.raise_for_status()
        batch = r.json()
        if not batch:
            return
        for order in batch:
            if float(order.get("total_refunded") or 0) > 0:
                yield order
        page += 1


def woo_intended_refund_minor(order):
    """What the shop manager actually asked WooCommerce to refund, in cents."""
    return round(float(order.get("total_refunded") or 0) * 100)


def get_charge_for_intent(intent_id):
    """The Stripe Charge behind a PaymentIntent or charge id, or None if not found."""
    if not intent_id:
        return None
    try:
        if intent_id.startswith("ch_"):
            return stripe.Charge.retrieve(intent_id)
        intent = stripe.PaymentIntent.retrieve(intent_id, expand=["latest_charge"])
        return intent.get("latest_charge")
    except stripe.error.InvalidRequestError:
        return None


def stripe_refunded_minor(charge):
    """What Stripe actually returned to the card for this charge, in cents."""
    return int((charge or {}).get("amount_refunded") or 0)


def decide(order, charge, woo_intended_minor, stripe_refunded_minor_value):
    """Pure decision: compare the intended refund to what Stripe actually moved.

    order: dict with at least "id" and "status".
    charge: dict-like Stripe Charge, or None if it could not be found.
    woo_intended_minor: cents the WooCommerce refund record says was refunded.
    stripe_refunded_minor_value: cents Stripe's charge.amount_refunded reports.

    Returns (action, reason, over_refund_minor).
    """
    if charge is None:
        return ("orphan", "no matching Stripe charge for this order", 0)
    if woo_intended_minor <= 0:
        return ("skip", "no refund recorded on this order", 0)
    gap = stripe_refunded_minor_value - woo_intended_minor
    if gap > TOLERANCE_MINOR:
        return ("overrefund", "Stripe refunded more than WooCommerce intended", gap)
    if gap < -TOLERANCE_MINOR:
        return ("underrefund", "Stripe refunded less than WooCommerce intended", gap)
    return ("ok", "Stripe refund matches the intended amount", 0)


def flag(order, reason, over_refund_minor):
    requests.post(
        f"{WOO_URL}/wp-json/wc/v3/orders/{order['id']}/notes",
        json={"note": (
            f"Refund check: {reason}. Stripe returned "
            f"{over_refund_minor / 100:.2f} more than the WooCommerce refund record shows. "
            f"This usually means the order's captured amount differs from Stripe's actual "
            f"amount_captured (a manual or partial capture). Review before refunding this "
            f"order again."
        )},
        auth=AUTH, timeout=30,
    ).raise_for_status()


def run():
    flagged = 0
    for order in recently_refunded_orders(LOOKBACK_HOURS):
        intent_id = intent_id_of(order)
        charge = get_charge_for_intent(intent_id)
        woo_intended = woo_intended_refund_minor(order)
        stripe_refunded = stripe_refunded_minor(charge)
        action, reason, over_refund_minor = decide(order, charge, woo_intended, stripe_refunded)
        if action == "orphan":
            log.warning("Order %s has a refund but no matching Stripe charge (%s)", order["id"], intent_id)
            continue
        if action in ("skip", "ok", "underrefund"):
            continue
        log.warning(
            "Order %s: %s. Woo intended %sc, Stripe refunded %sc. %s",
            order["id"], reason, woo_intended, stripe_refunded,
            "would flag" if DRY_RUN else "flagging",
        )
        if not DRY_RUN:
            flag(order, reason, over_refund_minor)
        flagged += 1
    log.info("Done. %d order(s) %s.", flagged, "to flag" if DRY_RUN else "flagged")


if __name__ == "__main__":
    run()

File: partial-refund-gives-back-everything/node/audit-refunds.js
/**
 * Catch WooCommerce partial refunds that actually returned the whole charge.
 *
 * On an order whose PaymentIntent was captured for less than the order total
 * (a manual capture, a phone order finished outside checkout, a split payment),
 * WooCommerce computes "amount left to refund" from the order total instead of
 * the real Stripe amount_captured. Ask for a small partial refund and the
 * gateway can send Stripe a refund with no amount, or an amount larger than
 * what is actually left, so Stripe refunds the entire remaining balance.
 *
 * This script compares, per order, the refund the shop manager intended
 * (the WooCommerce refund line item) against what Stripe actually refunded on
 * the matching charge. When Stripe refunded more than intended, it writes an
 * order note flagging the gap. It never asks Stripe to move money and never
 * edits amounts. It only reports. Read only by default. Run on a schedule.
 *
 * Guide: https://www.allanninal.dev/woocommerce/partial-refund-gives-back-everything/
 */
import Stripe from "stripe";
import { pathToFileURL } from "node:url";

const stripe = new Stripe(process.env.STRIPE_SECRET_KEY || "sk_test_dummy");
const WOO_URL = (process.env.WOO_STORE_URL || "https://example.com").replace(/\/$/, "");
const AUTH = "Basic " + Buffer.from(
  `${process.env.WOO_CONSUMER_KEY || "ck_dummy"}:${process.env.WOO_CONSUMER_SECRET || "cs_dummy"}`
).toString("base64");
const LOOKBACK_HOURS = Number(process.env.LOOKBACK_HOURS || 72);
const DRY_RUN = (process.env.DRY_RUN || "true").toLowerCase() === "true";

// Tolerance in cents for rounding noise between the two systems.
const TOLERANCE_MINOR = 1;

export function intentIdOf(order) {
  for (const meta of order.meta_data || []) {
    if (meta.key === "_stripe_intent_id" && meta.value) return meta.value;
  }
  const tid = order.transaction_id;
  return tid && (tid.startsWith("pi_") || tid.startsWith("ch_")) ? tid : null;
}

export function wooIntendedRefundMinor(order) {
  return Math.round(parseFloat(order.total_refunded || 0) * 100);
}

export function stripeRefundedMinor(charge) {
  return (charge && charge.amount_refunded) || 0;
}

/**
 * Pure decision: compare the intended refund to what Stripe actually moved.
 *
 * order: object with at least "id" and "status".
 * charge: Stripe Charge-shaped object, or null if it could not be found.
 * wooIntendedMinor: cents the WooCommerce refund record says was refunded.
 * stripeRefundedMinorValue: cents Stripe's charge.amount_refunded reports.
 *
 * Returns [action, reason, overRefundMinor].
 */
export function decide(order, charge, wooIntendedMinor, stripeRefundedMinorValue) {
  if (!charge) return ["orphan", "no matching Stripe charge for this order", 0];
  if (wooIntendedMinor <= 0) return ["skip", "no refund recorded on this order", 0];
  const gap = stripeRefundedMinorValue - wooIntendedMinor;
  if (gap > TOLERANCE_MINOR) return ["overrefund", "Stripe refunded more than WooCommerce intended", gap];
  if (gap < -TOLERANCE_MINOR) return ["underrefund", "Stripe refunded less than WooCommerce intended", gap];
  return ["ok", "Stripe refund matches the intended amount", 0];
}

async function woo(path, options = {}) {
  const res = await fetch(`${WOO_URL}/wp-json/wc/v3${path}`, {
    ...options,
    headers: { "Content-Type": "application/json", Authorization: AUTH, ...(options.headers || {}) },
  });
  if (!res.ok) throw new Error(`Woo ${path} returned ${res.status}`);
  return res.json();
}

async function getChargeForIntent(intentId) {
  if (!intentId) return null;
  try {
    if (intentId.startsWith("ch_")) return await stripe.charges.retrieve(intentId);
    const intent = await stripe.paymentIntents.retrieve(intentId, { expand: ["latest_charge"] });
    return intent.latest_charge || null;
  } catch {
    return null;
  }
}

async function* recentlyRefundedOrders(lookbackHours) {
  const since = new Date(Date.now() - lookbackHours * 3600 * 1000).toISOString();
  let page = 1;
  while (true) {
    const batch = await woo(
      `/orders?status=refunded,processing,completed&after=${since}&per_page=50&page=${page}`
    );
    if (!batch.length) return;
    for (const order of batch) {
      if (parseFloat(order.total_refunded || 0) > 0) yield order;
    }
    page++;
  }
}

async function flag(order, reason, overRefundMinor) {
  await woo(`/orders/${order.id}/notes`, {
    method: "POST",
    body: JSON.stringify({
      note: `Refund check: ${reason}. Stripe returned ${(overRefundMinor / 100).toFixed(2)} ` +
            `more than the WooCommerce refund record shows. This usually means the order's ` +
            `captured amount differs from Stripe's actual amount_captured (a manual or ` +
            `partial capture). Review before refunding this order again.`,
    }),
  });
}

export async function run() {
  let flagged = 0;
  for await (const order of recentlyRefundedOrders(LOOKBACK_HOURS)) {
    const intentId = intentIdOf(order);
    const charge = await getChargeForIntent(intentId);
    const wooIntended = wooIntendedRefundMinor(order);
    const stripeRefunded = stripeRefundedMinor(charge);
    const [action, reason, overRefundMinor] = decide(order, charge, wooIntended, stripeRefunded);
    if (action === "orphan") {
      console.warn(`Order ${order.id} has a refund but no matching Stripe charge (${intentId})`);
      continue;
    }
    if (action === "skip" || action === "ok" || action === "underrefund") continue;
    console.warn(
      `Order ${order.id}: ${reason}. Woo intended ${wooIntended}c, Stripe refunded ${stripeRefunded}c. ` +
      `${DRY_RUN ? "would flag" : "flagging"}`
    );
    if (!DRY_RUN) await flag(order, reason, overRefundMinor);
    flagged++;
  }
  console.log(`Done. ${flagged} order(s) ${DRY_RUN ? "to flag" : "flagged"}.`);
}

if (import.meta.url === pathToFileURL(process.argv[1]).href) {
  run().catch((err) => { console.error(err); process.exit(1); });
}
